k8srm-prototype/pod-spec-examples.yaml

```diff
@@ -0,0 +1,133 @@
+# containers, so there is no need to name the claim; the NICs will be added to
+# the pod-level sandbox.
+deviceClaims:
+- template:
+    claims:
+    - deviceClass: "intel-sriov-vf"
+      requests:
+        count: 4
+
+# Allocates four SR-IOV VFs from any vendor NIC, making sure all the VFs are on
+# the same PF this requires vendor & PF together to fully qualify a physical
+# card on a node.
+deviceClaims:
+- template:
+    claims:
+    - deviceType: "sriov-vf"
+      count: 4
+      matchAttributes:
+      - "vendor"
+      - "pf"
+
+# Allocates two A100 GPUs on the same PCI root complex. The name is used to
+# attach the associated claims to a container.
+deviceClaims:
+- template:
+    claims:
+    - name: gpus
+      deviceClass: "nvidia-a100"
+      count: 2
+      matchAttributes:
+      - pcie-root
+
+# Allocates two A100 GPUs on the same PCI root complex, but keeps them
+# separated in the API so they can be attached to two different containers.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu-one
+      deviceClass: "nvidia-a100"
+    - name: gpu-two
+      deviceClass: "nvidia-a100"
+
+# Allocates an SR-IOV PF and a an A100 from the same pcie-root. Name is
+# populated here for the PF just to show that it can be, but the expectation is
+# generally for network devices it is not needed. Note ANY vendor PF is fine
+# here; we use deviceType, not deviceClass.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu
+      deviceClass: "nvidia-a100"
+    - name: nic
+      deviceType: "sriov-pf"
+
+# Allocates either an A100, H100, or two L4s that are on the same pcie root.
+deviceClaims:
+- template:
+    claims:
+    - name: gpus
+      oneOf:
+      - deviceClass: "nvidia-a100"
+      - deviceClass: "nvidia-h100"
+      - deviceClass: "nvidia-l4"
+        count: 2
+        matchAttributes:
+        - pcie-root
+
+# Allocates and SR-IOV VF from any vendor and either an A100, H100, or two L4s,
+# all from the same NUMA node and PCIe root complex.
+deviceClaims:
+- template:
+    matchAttributes:
+    - numa
+    - pcie-root
+    claims:
+    - name: nic
+      deviceType: "sriov-vf"
+    - name: gpus
+      oneOf:
+      - deviceClass: "nvidia-a100"
+      - deviceClass: "nvidia-h100"
+      - deviceClass: "nvidia-l4"
+        count: 2
+
+# Allocates 4 VFs from 4 different PFs,
+# but all the same vendor
+deviceClaims:
+- template:
+  claims:
+  - name: nics
+    deviceType: "sriov-vf"
+    count: 4
+    matchAttributes:
+    - vendor
+    distinctAttributes: # maybe we should use affinity/antiaffinity terms?
+    - pf
+
+# Allocates a partial H100 in MPS mode, declaring a reservation of 8GB of
+# memory to regulate the number of clients that attach.
+deviceClaims:
+- template:
+  claims:
+  - name: gpus
+    deviceClass: "nvidia-h100-mps"
+    requests:
+      memory: 8Gi
```

Comment:
So in theory I could be allocated 16 devices of 0.5Gi each?  Is that actually useful?  If we make people jump thru hoops in order for us to DTRT, I am worried.  I feel like 1 should be the norm, and allowing more than one is the exception?

Reply:
My argument for MPS (and sharing in general) is that is should be user directed. You allocate a claim asking for a certain number of phsoycal GPUs (or MIg devices) in a certain sharing mode (MPS or timeslicing) with a set of sharing-mode-specific parameters, and then you, as the user, decide which workloads to point at that claim to share access to the GPUs in a controlled way.

Reply:
Kevin: Whether explicit sharing or not, do you think it is "useful" to allocate 16 * 0.5Gi to satisfy a request for 8Gi?  Surely there's some impact on the consumer(s), and some overhead and boundary size below which it's useless?  E.g. 32 * 0.25Gi or 64 * 0.125Gi seems like a stretch, and (to be silly) 1024 * 8Mi seems ludicrous.

> you, as the user, decide which workloads to point at that claim

This really demands that the DeviceClaim be a distinct object (lifecycle != pod), like VolumeClaim.  The ugly part of that is, like VolumeClaim, you can strand the device on a machine which has insufficient CPU or memory to schedule the consumer pod(s).

Reply:
So, this example was about a PARTIAL device. Not multiple devices aggregated to satisfy a request.

We may want to default to a claim being satisfied by a single device, and allow for a specific number of devices if desired. And limit it to that for 1.31. 

Later, if there is demand, we can offer a way to ask for a flexible number of devices to meet a resource allocation request. 

The one exception is the need (?) for an "all devices meeting the constraints expression" which we would use for the "management" claims. Or perhaps we make that the ONLY behavior of management claims?

Reply:
I would be OK assuming that management claims imply access to *all* devices matching a specific selection criteria (rather than a certain number of them).